### ISSUE TITLE
Native methods for filtering mode data

### ISSUE THREAD

**Issue by tomflexcompute** [flexcompute/tidy3d#1391]
A user reported that they spend a lot of time on either filtering modes manually or implementing their own filters. What could help them:
1.	a flag to automatically filter out modes that are marked as "not decayed at the boundary".
2.	filter by the loss, i.e. provide a threshold for how much of loss is allowed and remove all modes above that.
3.	filter by mode volume: similarly, provide a threshold value for mode volume.

And thanks @dbochkov-flexcompute for collecting the information from the user.

**Comment by momchil-flex** [comment]
From discussion, few possible ways:
1. Define frontend post-processing methods that can be e.g. `ModeSolverData.filter(filter_what) -> FilteredModeSolverData`, where `FilteredModeSolverData` also keeps the indexing into the original mode solver data so that a user can e.g. select a mode for injection in a `ModeSource`. Pros: easier to implement and use directly, including different filtering in post-processing. Cons: a bit more confusing how to find modes in the original data; harder to incorporate in the GUI post-run viz.
2. Extend the filtering definition in `ModeSpec` (currently only `filter_pol`, convert to a `filter` dictionary or a `FilterSpec` that defines various filtering options) such that the produced modes are directly filtered as requested. Pros: better-defined, smaller chance to mistake definitions in a mode source or mode monitor, easier to incorporate in GUI. Cons: less flexible, requires a new mode solver for every filtering spec.
3. A combination of the two? But it can make things confusing.

I'm kinda leaning towards option 2. still - @dbochkov-flexcompute @tomflexcompute @xin-flex what do you think?

**Comment by dbochkov-flexcompute** [comment]
yeah, I agree, I think option 2 is good. Are you thinking of filtering as elimination of modes (threshold for loss, threshold for volume, etc) or reordering of them (order by loss, order by volume, put non-decayed modes to the end, etc)?

**Comment by momchil-flex** [comment]
> yeah, I agree, I think option 2 is good. Are you thinking of filtering as elimination of modes (threshold for loss, threshold for volume, etc) or reordering of them (order by loss, order by volume, put non-decayed modes to the end, etc)?

This is the biggest question... Filtering in the sense of elimination probably makes more sense from the point of view of the user. The biggest issue though is that in our current workflow if a mode source is created with a filtering spec that reduces the number of modes, and `mode_index` is set to a higher value than what's available after elimination, we need to error. But generally so far we have no workflow that errors by design, we try to catch everything in advance and if an error happens it's something we need to look at. Because of that, it's tempting to go with rearranging. On the other hand, again from user experience, if the `mode_index` is pointing to some of the filtered out modes, it's likely that something's wrong and the user probably does want the simulation aborted. I'm just not sure how to set it up though.

**Comment by yuanshen-flexcompute** [comment]
Hi @momchil-flex , just digging this issue back up again as this feature (sorting/filtering modes by numerical criteria) can also be useful for RF Wave Ports. This came up in discussion with RF team: @weiliangjin2021 @dbochkov-flexcompute @dmarek-flex 

To side-step the missing modes issue, we can just sort instead of filter. The easiest (?) implementation would be to sort by a single criteria (user-chosen mode property by increasing/decreasing/distance to user-specified value). If mode tracking is enabled, the modes are sorted only at the tracking start frequency (to preserve mode order). If mode tracking is disabled, then they are sorted at every frequency.

**Comment by momchil-flex** [comment]
Nice, I like this suggestion! The current behavior is exactly that, with the sorting being by `n_eff` and the only freedom that the user has is to set the `target_neff`. Actually even that does not directly affect the sorting, just the modes which are computed, but eventually they are still sorted by decreasing `n_eff`, not by their "closeness" to `target_neff`.

So yeah I think we can introduce a new `ModeSortSpec` that can define which value to sort by, as well as a reference value. What should supported values be? I guess in principle all of these could be nice in some use cases? Is there anything missing?

<img width="1029" height="178" alt="Image" src="https://github.com/user-attachments/assets/ce290f21-8a22-4f3f-8424-5f44306190af" />

**Comment by yuanshen-flexcompute** [comment]
It probably makes sense to support all the existing values in `ModeData.modes_info`. 

A nice-to-have feature would be to allow arbitrary user-defined functions for sorting, like behaviour of Python `sorted()` function with the `key` argument. 

Eventually it might be nice to sort by more than one criteria, but that might complicate things too much.

**Comment by yaugenst-flex** [comment]
Closing — native mode filtering was implemented across three PRs: mode sorting in PR #2869 (FXC-3547), drop filtered modes in PR #3055 (FXC-5121), and updated defaults in PR #3144 (FXC-4643).
